### PR TITLE
Reject shared_embeddings=true for untied models in model builder

### LINKED ISSUE #2306: [Bug] Model builder silently ties an untied model: shared_embeddings=true on gpt-oss-20b produces gibberish
## Summary

`openai/gpt-oss-20b` uses **untied** input/output embeddings (`tie_word_embeddings: false`, with distinct `model.embed_tokens.weight` and `lm_head.weight` tensors). Building it to ONNX with `shared_embeddings=true` tells the model builder to tie those two weights anyway. The builder honors the flag **without any error or warning**, so the exported model reads token embeddings from the wrong weight matrix and emits **completely garbled output** for any prompt: greedy generation is incoherent and the logits are finite-but-flat (per-token log-prob ≈ `ln(1/vocab)` ≈ `-12.2` for `vocab = 201,088`), so accuracy on multiple-choice benchmarks (via [lm-evaluation-harness](https://github.com/EleutherAI/lm-evaluation-harness)) collapses to the random-guess floor (`arc_challenge acc ≈ 0.23`, i.e. ~1/4).

The same `builder.py` flags produce a **healthy, coherent** model for `Qwen3-4B`, `Llama-3.2-3B-Instruct` and `Phi-4-Mini-Instruct` - and that is fully consistent with the root cause, because those models genuinely tie their embeddings.

## Environment

- Model: `openai/gpt-oss-20b` (`gpt_oss`, 24 layers, hidden 2880, vocab 201088, `tie_word_embeddings: false`)
- Backend: ONNX Runtime GenAI, WebGPU execution provider (also reproduces on the CPU)
- Precision: int4 (RTN, symmetric)
- Tool: `onnxruntime-genai\src\python\py\models\builder.py` (onnxruntime-genai model builder)
- `onnxruntime-genai` version: `87047e6`
- `onnxruntime` version: `da90494`
- OS: Windows 11

### 1. Convert the model

```bash
python builder.py -m openai/gpt-oss-20b -o webgpu-gpt-oss-20b-onnx -c cache-dir -p int4 -e webgpu --extra_options shared_embeddings=true int4_algo_config=rtn_last int4_is_symmetric=true enable_webgpu_graph=true prune_lm_head=true hf_remote=false
```

### 2. Run the model benchmark / generation

`model_benchmark.exe` is the benchmarking/generation tool shipped with
onnxruntime-genai (not a custom script):

```bash
model_benchmark.exe -i webgpu-gpt-oss-20b-onnx --prompt "What is ONNXRuntime?" -g 128 -r 1 -v
```

## Actual Behavior

The output is **all garbled / gibberish**:

```
[PROMPT BEGIN]What is ONNXRuntime?[PROMPT END]
[OUTPUT BEGIN]What is ONNXRuntime?**: ONNX Runtime µÿ»Σ╕ÇΣ╕¬Θ½ÿµÇºΦâ╜µ╖▒σ║ªσ¡ªΣ╣áµÄ¿τÉåσÆîµÄ¿τÉåµ£ìσèí∩╝îΘÇéτö¿Σ║ÄσñÜτºìτí¼Σ╗╢τÄ»σóâ∩╝îσ╣╢σÅ»Φ╖¿σñÜΦ»¡Φ¿ÇΣ╜┐τö¿πÇéσ«âΣ╕ôΘù¿Φó½Φ«╛Φ«íΣ╕║σÅ»µë⌐σ▒òσÆîµ¿íσ¥ùσîû∩╝îµö»µîüσñÜτºìµÄ¿τÉåσèáΘÇƒµèÇµ£»∩╝îΣ╛ïσªé GPUπÇüFPGA Σ╗ÑσÅèσà╢Σ╗ûΣ╕ôτö¿σèáΘÇƒσìí∩╝îσÅ»Σ╗ÑΣ╝ÿσîûµ╖▒σ║ªσ¡ªΣ╣áµ¿íσ₧ïτÜäµÄ¿τÉåΦ┐ÉΦíîπÇé

2. **ONNX Runtime τÜäσ╖ÑΣ╜£σÄƒτÉå**: σ«âΘÇÜΦ┐çσ░å ONNX Θççτö¿τÜäµ╖▒σ║ªσ¡ªΣ╣áµ¿íσ₧ïΣ╕Äτí¼Σ╗╢σèáΘÇƒσÖ¿πÇüσåàµá╕πÇüσåàσ¡ÿΣ╝ÿσîûσ╖Ñσà╖τ¡ëσñÜτºìµèÇµ£»τ╗ôσÉêΣ╜┐τö¿∩╝îΣ╗ÑµÅÉΘ½ÿµÄ¿τÉåΘÇƒσ║ª[OUTPUT END]
```

## Expected Behavior

Coherent text, e.g.:

```
[PROMPT BEGIN]What is ONNXRuntime?[PROMPT END]
[OUTPUT BEGIN]What is ONNXRuntime?

ONNX Runtime is an open-source cross-platform runtime for executing machine learning models in the Open Neural Network Exchange (ONNX) format. It is designed to optimize the performance of ONNX models across a variety of hardware platforms, including CPUs, GPUs, FPGAs, and specialized ML accelerators.

ONNX Runtime was developed by Microsoft in collaboration with the ONNX community. It aims to provide a consistent and efficient way to deploy ONNX models across different hardware platforms and software frameworks, including TensorFlow, PyTorch, and scikit-learn.

ONNX Runtime provides two main components:

1) an inference engine: this takes an[OUTPUT END]
```

## Root Cause

gpt-oss-20b has **untied** embeddings:

- `config.json` sets `"tie_word_embeddings": false`.
- The checkpoint ships **two distinct tensors**, `model.embed_tokens.weight` and  `lm_head.weight` - they are not the same matrix.

`shared_embeddings=true` instructs the builder to tie the input embedding to the LM-head weight. In `make_tied_embeddings_init` (`src\python\py\models\builders\base.py`) the **explicit flag overrides** the model's `tie_word_embeddings=false`, and `make_embedding` then builds the token-embedding `Gather` from the LM-head weight (`lm_head.MatMul.weight`) instead of `model.embed_tokens.weight`. For a genuinely tied model those two are identical, so the substitution is harmless; for gpt-oss they differ, so **every token embedding is read from the wrong matrix** and the network produces flat/garbled logits from layer 0. The builder applies this override with no error or warning - that silent acceptance is the bug.

### PR DESCRIPTION
**Pull request by haoxli**
The model builder silently honored shared_embeddings=true even when the model config sets tie_word_embeddings=false. For untied models (e.g. gpt-oss-20b and DeepSeek-R1-Distill-Qwen-1.5B), the token embedding and LM head are separate weights, so tying them made the embedding read from the wrong matrix and exported a broken model: gibberish generation, finite-but-flat logits, and near-random benchmark accuracy -- with no error or warning.

Raises a clear ValueError in create_model() when shared_embeddings=true for a model whose config has tie_word_embeddings=false, before any conversion work runs. Genuinely tied models (tie_word_embeddings=true) are unaffected.

Fixes #2306

### CODE REVIEW

File: src/python/py/models/builder.py [hunk 1]
```diff
@@ -256,6 +256,20 @@ def create_model(
         )
         config.update(peft_config.__dict__)
 
+    # Weight sharing (shared_embeddings=true) reuses a single matrix for both the input
+    # embedding and the LM head. That is only valid when the model actually ties them. For an
+    # untied model (config.tie_word_embeddings is False) the token embedding and LM head are
+    # distinct weights, so tying them would make the embedding read from the wrong matrix and
+    # silently export a broken model (e.g. gpt-oss-20b generating gibberish). Reject the
+    # combination up front instead of producing a corrupt model.
+    if extra_options.get("shared_embeddings", False) and getattr(config, "tie_word_embeddings", None) is False:
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builder.py:265]: Can you move this check inside `check_extra_options` instead?

Comment (haoxli) [reply src/python/py/models/builder.py:265]: The check needs the loaded model config (`config.tie_word_embeddings`). `check_extra_options` runs at argument-parse time before the config exists, so it has nothing to check against. Even if we pass the config and check it inside `check_extra_options`, we also need to call `check_extra_options` again in `create_model`, which will crash on already-normalized values.

Comment (kunal-vaishnavi) [reply src/python/py/models/builder.py:265]: By the time `create_model` is called, we should already have confirmed all of the values in the `extra_options` (whether called via CLI or via API). I have opened [this PR](https://github.com/microsoft/onnxruntime-genai/pull/2312) to make this work.

**Comment by kunal-vaishnavi** [comment]
Closing in favor of https://github.com/microsoft/onnxruntime-genai/pull/2312. I cherry-picked your changes into the new PR.
